src/main/kotlin/org/rust/ide/intentions/UnwrapToMatchIntention.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.intentions
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.psi.ext.parentDotExpr
+import org.rust.lang.core.psi.ext.receiver
+import org.rust.lang.core.resolve.knownItems
+import org.rust.lang.core.types.ty.TyAdt
+import org.rust.lang.core.types.type
+
+class UnwrapToMatchIntention: RsElementBaseIntentionAction<UnwrapToMatchIntention.Context>() {
+    override fun getText() = "Replace .unwrap() with match"
+    override fun getFamilyName() = text
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val methodCall = element.ancestorOrSelf<RsMethodCall>() ?: return null
+        val itemType = (methodCall.receiver.type as? TyAdt)?.item as? RsEnumItem ?: return null
+        val isOption = isStdOptionOrResult(itemType)
+
+        if (methodCall.referenceName == "unwrap" &&
+            methodCall.typeArgumentList == null &&
+            methodCall.valueArgumentList.exprList.isEmpty() &&
+            isOption != null) {
+            return Context(methodCall, isOption)
+        }
+        return null
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val (methodCall, isOption) = ctx
+        val generatedCode = buildString {
+            append("match ")
+            append(methodCall.receiver.text)
+            append(" {")
+            append(if (isOption) "Some" else "Ok")
+            append("(x) => x,")
+            append(if (isOption) "None" else "Err(_)")
+            append(" => unimplemented!(),")
+            append("}")
+        }
+
+        val matchExpression = RsPsiFactory(project).createExpression(generatedCode) as RsMatchExpr
+        methodCall.parentDotExpr.replace(matchExpression)
+    }
+
+    private fun isStdOptionOrResult(item: RsEnumItem): Boolean? {
```

Comment:
I'd recommend to crate a private enum like `ResultKind` with 2 variants and return `ResultKind?` here. Also it's possible to add properties like `okVariant` (`Some`/`Ok`) and `errorVariant` to that enum.